### PR TITLE
#165299388 User can Create Articles 

### PR DESCRIPTION
**Pull request by RachelleMaina**
### What does this PR do?
 -  Enables users to create articles if they are logged in

### Description
 - adds a 3rd party editor
 -  adds utils to upload image to cloudinary
 - adds a dropdown on homepage
 - Adds protected routing
 -  adds tests

### How has this been tested?
 - [x] Unit tests

### How to manually test this:
- Clone the repository `https://github.com/andela/ah-legion-frontend`
- Checkout the branch `ft-create-article-165299388`
- Run `npm install` to install dependecies
-  Run `npm start' to fire up the dev server
-  Go to a browser, login and visit the  url ie. `/article/create` and `/articles/drafts`
- Protected routes are those routes that cannot be accessed without authentication. e.g In the routes file, do this to protect the route:
 `<ProtectedRoute path="/articles/drafts" component={DraftsView} />`
If a user is not logged in, they will be redirected to the homepage.

### PT
[#165299388](https://www.pivotaltracker.com/story/show/165299388)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,5 +26,9 @@ yarn-debug.log*
 yarn-error.log*
 
 package-lock.json
+<<<<<<< HEAD
 
 /.vscode
+=======
+.env
+>>>>>>> feature(article list):user can see list of articles
```
Comment (abulojoshua1) [inline .gitignore]: @RachelleMaina, fix merge conflicts fully.

File: src/axiosConfig.js [hunk 2]
```diff
@@ -1,7 +1,6 @@
 import axios from 'axios';
 
 const { REACT_APP_BASE_URL } = process.env;
-
```
Comment (abulojoshua1) [inline src/axiosConfig.js]: @RachelleMaina, is this change necessary?

File: src/components/CreateArticle.jsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  onSubmit,
+  onEditorStateChange,
+  handleChange,
+  tags,
+  draftStatus,
+  handleDelete,
+  handleAddition,
+  handleDrag,
+  onPublish,
+  show,
```
Comment (abulojoshua1) [inline src/components/CreateArticle.jsx]: @RachelleMaina, it is better to parse props and destructure instead, it makes it more maintainable.

File: src/containers/CreateArticleView.jsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  state = {
+    title: '',
+    description: '',
+    editorState: EditorState.createEmpty(),
+    tagsInputValue: '',
+    tags: [],
+    count: 0,
+    draftStatus: '',
+    show: false,
+  };
```
Comment (abulojoshua1) [inline src/containers/CreateArticleView.jsx:52]: @RachelleMaina, use a constructor for state to make the app uniform.

File: src/containers/CreateArticleView.jsx [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    this.setState({
+      [e.target.id]: e.target.value,
+      count: count + 1,
+    },
+    () => {
+      this.create();
+    });
+  };
+
+  handleDelete = (i) => {
```
Comment (abulojoshua1) [inline src/containers/CreateArticleView.jsx]: @RachelleMaina, use a variable name line index and not `i`

File: src/css/App.css [hunk 6]
```diff
@@ -712,3 +713,196 @@ a {
...
+}
+div.ReactTags__suggestions li mark {
+    text-decoration: underline;
+    background: none;
+    font-weight: 600;
+}
+div.ReactTags__suggestions ul li.ReactTags__activeSuggestion {
+    background: #b7cfe0;
+    cursor: pointer;
+}
```
Comment (abulojoshua1) [inline src/css/App.css:913]: @RachelleMaina, add EOF here.

File: src/tests/components/articles.test.jsx [hunk 7]
```diff
@@ -31,3 +34,47 @@ describe('AllArticles ', () => {
...
+        handleChange={handleChange}
+        onEditorStateChange={onEditorStateChange}
+        tags={tags}
+        uploadImageCallBack={uploadImageCallBack}
+        draftStatus={draftStatus}
+        handleDelete={handleDelete}
+        handleAddition={handleAddition}
+        handleDrag={handleDrag}
+        onPublish={onPublish}
+        show={show}
```
Comment (abulojoshua1) [inline src/tests/components/articles.test.jsx]: @RachelleMaina, create a `const props = {    const onSubmit = jest.fn();
    const handleChange = jest.fn();
    const onEditorStateChange = jest.fn();
    const tags = [{ id: 1, test: 'tags' }];
    const draftStatus = 'saving';
    const handleDelete = jest.fn();
    const handleAddition = jest.fn();
    const handleDrag = jest.fn();
    const onPublish = jest.fn();
    const show = false;
}` and spread it in `<CreateArticle />` using `<CreateArticle {...props} />`

**Review by abulojoshua1 — CHANGES_REQUESTED** [review]
Looks good, just make a few changes while still in progress.

File: src/axiosConfig.js (hunk 2 above)
Comment (RachelleMaina) [reply src/axiosConfig.js]: Not really. I have corrected this.

**Comment by RachelleMaina** [comment]
> Looks good, just make a few changes while still in progress.

I have made the changes. Just 3 comments from hound which I haven't found a way to solve them.

**Review by MandelaK — CHANGES_REQUESTED** [review]
Please don't track `package-lock.json`

**Comment by RachelleMaina** [comment]
> Please don't track `package-lock.json`

I have corrected this

**Review by MandelaK — APPROVED** [review]

**Comment by MandelaK** [comment]
Could you please add a brief documentation for how to use PrivateRoutes?

File: src/routes/index.jsx [hunk 8]
```diff
@@ -13,6 +17,9 @@ const Routes = () => (
       <Route exact path="/" component={HomeView} />
       <Route path="/profile" component={UserProfile} />
       <Route path="/reset-password/:resetToken" component={ResetPassword} />
+      <Route path="/login" component={Login} />
+      <ProtectedRoute path="/article/create" component={CreateArticleView} />
+      <ProtectedRoute path="/articles/drafts" component={DraftsView} />
```
Comment (abulojoshua1) [inline src/routes/index.jsx]: @RachelleMaina, protect these routes from being accessed by a user that is not logged in, connect this component with redux, pick the state that determines a user's login status, make a condition that renders that route based on that state.

**Review by abulojoshua1 — APPROVED** [review]
LGTM
